File: scripts/wan22_vae_encode_parity.py
#!/usr/bin/env python
"""Wan2.2 VAE encoder parity test — Python reference.

Loads the same random input that the Rust parity binary generated,
encodes it with the Python Wan2.2 VAE, and compares the output.

Usage:
    # First run the Rust binary to generate the input:
    #   cargo run --release --bin wan22_vae_encode_parity
    # Then run this:
    python scripts/wan22_vae_encode_parity.py
"""
from __future__ import annotations

import os
import sys
import time

import torch
import torch.cuda.amp as amp
import numpy as np
from safetensors.torch import save_file, load_file

# Paths
VAE_PTH = os.environ.get(
    "WAN_VAE_PTH",
    "/home/alex/.serenity/models/checkpoints/Wan2.2-TI2V-5B/Wan2.2_VAE.pth",
)
OUTPUT_DIR = "/home/alex/EriDiffusion/inference-flame/output"
INPUT_PATH = f"{OUTPUT_DIR}/wan22_vae_encode_input.safetensors"
RUST_OUTPUT_PATH = f"{OUTPUT_DIR}/wan22_vae_encode_output.safetensors"
PY_OUTPUT_PATH = f"{OUTPUT_DIR}/wan22_vae_encode_output_python.safetensors"

# Add Wan2.2 to path
sys.path.insert(0, "/home/alex/Wan2.2")
from wan.modules.vae2_2 import Wan2_2_VAE


def stats(name: str, t: torch.Tensor):
    t_f = t.float()
    print(
        f"  [{name}] shape={tuple(t.shape)} "
        f"mean={t_f.mean():.4f} |mean|={t_f.abs().mean():.4f} "
        f"range=[{t_f.min():.4f}, {t_f.max():.4f}] "
        f"nan={t_f.isnan().sum().item()} inf={t_f.isinf().sum().item()}"
    )


def main() -> int:
    device = "cuda"
    dtype = torch.bfloat16

    print("=== Wan2.2 VAE Encoder Parity Test (Python) ===\n")

    # --- Load input from Rust ---
    if not os.path.exists(INPUT_PATH):
        print(f"ERROR: Input file not found: {INPUT_PATH}")
        print("Run the Rust binary first: cargo run --release --bin wan22_vae_encode_parity")
        return 1

    loaded = load_file(INPUT_PATH, device="cpu")
    video = loaded["video"].to(device=device, dtype=dtype)
    stats("input video", video)
    print()

    # --- Load VAE ---
    print("--- Loading Wan2.2 VAE ---")
    t0 = time.time()
    vae = Wan2_2_VAE(
        z_dim=48,
        c_dim=160,
        vae_pth=VAE_PTH,
        dim_mult=[1, 2, 4, 4],
        temperal_downsample=[False, True, True],
        dtype=dtype,
        device=device,
    )
    print(f"  Loaded in {time.time() - t0:.1f}s\n")

    # --- Encode ---
    print("--- Encoding ---")
    t1 = time.time()
    with torch.no_grad(), amp.autocast(dtype=dtype):
        # Use the model's encode method which handles patchify + normalize
        latent = vae.model.encode(video, vae.scale).float()
    print(f"  Encode: {time.time() - t1:.1f}s")
    stats("output latent", latent)

    # Save Python output
    save_file({"latent": latent.cpu()}, PY_OUTPUT_PATH)
    print(f"  Saved to {PY_OUTPUT_PATH}\n")

    # --- Compare with Rust if available ---
    if os.path.exists(RUST_OUTPUT_PATH):
        print("--- Comparing Rust vs Python ---")
        rust_data = load_file(RUST_OUTPUT_PATH, device="cpu")
        rust_latent = rust_data["latent"]
        py_latent = latent.cpu().float()

        stats("rust latent", rust_latent)
        stats("python latent", py_latent)

        diff = (rust_latent - py_latent).abs()
        print(f"\n  Absolute diff: mean={diff.mean():.6f} max={diff.max():.6f}")
        print(f"  Relative diff: mean={(diff / (py_latent.abs() + 1e-8)).mean():.6f}")

        # Cosine similarity per-sample
        rust_flat = rust_latent.reshape(1, -1)
        py_flat = py_latent.reshape(1, -1)
        cos_sim = torch.nn.functional.cosine_similarity(rust_flat, py_flat)
        print(f"  Cosine similarity: {cos_sim.item():.6f}")

        if diff.max() < 0.1:
            print("\n  PASS: Outputs are close (max diff < 0.1)")
        elif diff.max() < 1.0:
            print("\n  WARN: Outputs have moderate differences (max diff < 1.0)")
        else:
            print("\n  FAIL: Large differences detected")
    else:
        print(f"Rust output not found at {RUST_OUTPUT_PATH}")
        print("Run Rust binary first for comparison.")

    print("\n=== Done ===")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/bin/wan22_vae_encode_parity.rs
//! Wan2.2 VAE encoder parity test.
//!
//! Loads the Wan2.2 VAE encoder, encodes a small random video tensor,
//! saves the output latent to safetensors for comparison with Python.
//!
//! Usage:
//!     cargo run --release --bin wan22_vae_encode_parity
//!
//! Set WAN_VAE_DBG=1 for per-layer activation stats.

use std::collections::HashMap;
use std::path::Path;
use std::time::Instant;

use flame_core::{global_cuda_device, DType, Shape, Tensor};
use flame_core::serialization;

use inference_flame::models::wan::Wan22VaeEncoder;

const VAE_PATH: &str = "/home/alex/.serenity/models/vaes/wan2.2_vae.safetensors";
const OUTPUT_DIR: &str = "/home/alex/EriDiffusion/inference-flame/output";

fn tensor_stats(name: &str, t: &Tensor) -> anyhow::Result<()> {
    let data = t.to_dtype(DType::F32)?.to_vec1::<f32>()?;
    let n = data.len();
    let nan = data.iter().filter(|v| v.is_nan()).count();
    let inf = data.iter().filter(|v| v.is_infinite()).count();
    let mean: f32 = data.iter().copied().sum::<f32>() / (n as f32);
    let abs_mean: f32 = data.iter().map(|v| v.abs()).sum::<f32>() / (n as f32);
    let (mut lo, mut hi) = (f32::INFINITY, f32::NEG_INFINITY);
    for &v in &data {
        if v.is_finite() {
            if v < lo { lo = v; }
            if v > hi { hi = v; }
        }
    }
    println!(
        "  [{name}] shape={:?}  n={n}  nan={nan}  inf={inf}  mean={mean:.4}  |mean|={abs_mean:.4}  range=[{lo:.4}, {hi:.4}]",
        t.shape().dims(),
    );
    Ok(())
}

fn main() -> anyhow::Result<()> {
    env_logger::init();
    let device = global_cuda_device();

    println!("=== Wan2.2 VAE Encoder Parity Test ===\n");

    // --- Create random input video [1, 3, 5, 256, 256] in [-1, 1] ---
    // Use a fixed seed for reproducibility by generating from a known pattern.
    let (b, c, t, h, w) = (1, 3, 5, 256, 256);
    let n = b * c * t * h * w;

    // Generate pseudo-random BF16 values in [-1, 1] using a simple LCG
    let mut rng_state: u64 = 42;
    let mut vals = Vec::with_capacity(n);
    for _ in 0..n {
        rng_state = rng_state.wrapping_mul(6364136223846793005).wrapping_add(1);
        let u = (rng_state >> 33) as f32 / (1u64 << 31) as f32; // [0, 1)
        vals.push(u * 2.0 - 1.0); // [-1, 1)
    }

    let video = Tensor::from_vec(
        vals,
        Shape::from_dims(&[b, c, t, h, w]),
        device.clone(),
    )?.to_dtype(DType::BF16)?;
    tensor_stats("input video", &video)?;

    // Save input for Python comparison
    let input_path = format!("{OUTPUT_DIR}/wan22_vae_encode_input.safetensors");
    std::fs::create_dir_all(OUTPUT_DIR)?;
    let mut input_tensors = HashMap::new();
    input_tensors.insert("video".to_string(), video.to_dtype(DType::F32)?);
    serialization::save_file(&input_tensors, &input_path)?;
    println!("  Saved input to {input_path}\n");

    // --- Load encoder ---
    println!("--- Loading Wan22 VAE Encoder ---");
    let t0 = Instant::now();
    let encoder = Wan22VaeEncoder::load(Path::new(VAE_PATH), &device)?;
    println!("  Loaded in {:.1}s\n", t0.elapsed().as_secs_f32());

    // --- Encode ---
    println!("--- Encoding ---");
    let t1 = Instant::now();
    let latent = encoder.encode(&video)?;
    println!("  Encode: {:.1}s", t1.elapsed().as_secs_f32());
    tensor_stats("output latent", &latent)?;

    // Save output
    let output_path = format!("{OUTPUT_DIR}/wan22_vae_encode_output.safetensors");
    let mut output_tensors = HashMap::new();
    output_tensors.insert("latent".to_string(), latent.to_dtype(DType::F32)?);
    serialization::save_file(&output_tensors, &output_path)?;
    println!("  Saved output to {output_path}\n");

    println!("=== Done ===");
    println!("Run the Python comparison script to verify parity:");
    println!("  python scripts/wan22_vae_encode_parity.py");

    Ok(())
}
